packages/user/Tokens/service/src/tables.rs

```diff
@@ -727,4 +745,185 @@ pub mod tables {
             upt.erase(&(debitor, token_id, shared_bal_id));
         }
     }
+
+    #[table(name = "SubAccountTable", index = 8)]
+    #[derive(Serialize, Deserialize, ToSchema, Fracpack, Debug, SimpleObject)]
+    pub struct SubAccount {
+        pub account: AccountNumber,
+        pub sub_account: String,
+        #[graphql(skip)]
+        pub id: u64,
+    }
+    impl SubAccount {
+        #[primary_key]
+        fn pk(&self) -> (AccountNumber, String) {
+            (self.account, self.sub_account.clone())
+        }
+
+        #[secondary_key(1)]
+        fn by_id(&self) -> u64 {
+            self.id
+        }
+    }
+    impl SubAccount {
+        fn new(owner: AccountNumber, sub_account: String) -> Self {
+            Self {
+                account: owner,
+                sub_account,
+                id: InitRow::next_subaccount_id(),
+            }
+        }
+
+        pub fn get(owner: AccountNumber, sub_account: String) -> Option<Self> {
+            SubAccountTable::read()
+                .get_index_pk()
+                .get(&(owner, sub_account))
+        }
+
+        pub fn get_or_add(owner: AccountNumber, sub_account: String) -> Self {
+            Self::get(owner, sub_account.clone()).unwrap_or_else(|| {
+                let mut record = Self::new(owner, sub_account);
+                record.save();
+                record
+            })
+        }
+
+        pub fn get_assert(owner: AccountNumber, sub_account: String) -> Self {
+            check_some(
+                Self::get(owner, sub_account.clone()),
+                &format!("{} has no sub-account '{}'", owner, sub_account),
+            )
+        }
+
+        pub fn get_by_id(id: u64) -> Option<Self> {
+            SubAccountTable::read().get_index_by_id().get(&id)
+        }
+
+        fn save(&mut self) {
+            let table = SubAccountTable::new();
+            table.put(&self).unwrap();
+        }
+
+        pub fn delete(&mut self) {
+            let sub_balances: Vec<_> = SubAccountBalanceTable::read()
+                .get_index_pk()
+                .range((self.id, 0)..(self.id, u32::MAX))
+                .collect();
+
+            for mut sub_balance in sub_balances {
+                if sub_balance.balance.value > 0 {
+                    Balance::get_or_new(self.account, sub_balance.token_id)
+                        .add_balance(sub_balance.balance);
+                }
+                sub_balance.delete();
```

Comment:
This calls `table::read_write` for every sub_balance rather than just the once if we initialised it outside of the loop. Are we concerned about this or not a big deal?

Reply:
It should be moved outside the loop

Reply:
I fixed this particular instance but there's actually a ton of these kinds of inefficiencies. These interfaces would be better in general if we were passing handles around instead of fetching new handles all the time. We can clean these up later when we are optimizing for more efficient cpu usage